etc/pr_check/scan.go

```diff
@@ -0,0 +1,190 @@
+		// [start, end] is close to now, this will be "desc" (most recent PRs
+		// first). If the interval [start, end] is close to 'earliestPR', this will
+		// be "asc" (oldest PRs first).
+		direction string
+
+		// startGap is, very roughly, the amount of time before 'start' and the
+		// amount of time after 'end'. Either or both durations may be negative--I
+		// made up 'earliestPR', and 'end' is defined to be tomorrow by default..
+		startGap, endGap time.Duration
+
+		// Which page of GitHub results is being read
+		page = 1
+
+		result []*github.PullRequest
+	)
+
+	// Initialize gaps to figure out if [start, end] is "old". Both gaps may be
+	// negative (start before 'earliestPR', end in future. main() initializes
+	// 'end' to be in the future, so 'endGap' is actually usually negative), but
+	// the logic should still be correct.
+	startGap = start.Sub(earliestPR)
+	endGap = now.Sub(end)
+
+	// Initialize 'direction' and 'stopFn'
+	direction = "desc"
+	if startGap < endGap {
+		direction = "asc"
+	}
+
+	// Read data, page by page
+forEachPage:
+	for {
+		fmt.Printf("Getting PRs, page %d...", page)
+		nextPrs, resp, err := client.PullRequests.List(context.Background(),
+			"pachyderm", "pachyderm",
+			&github.PullRequestListOptions{
+				Direction: direction,
+				ListOptions: github.ListOptions{
+					Page: page,
+				},
+			})
+		if err != nil {
+			io.Copy(os.Stdout, resp.Body) // only way to see detailed err text
+			panic("could not request open Pachyderm PRs from GitHub: " + err.Error())
+		}
+		fmt.Printf("of %d. (Got %d PRs %s)\n", resp.LastPage, len(nextPrs), prsSummaryStr(nextPrs))
+		for _, pr := range nextPrs {
+			beforeRange := (direction == "asc" && pr.CreatedAt.GetTime().Before(start)) ||
+				(direction == "desc" && pr.CreatedAt.GetTime().After(end))
+			if beforeRange {
+				continue // skip PRs that are before the beginning of [start, end]
+			}
+			afterRange := (direction == "asc" && pr.CreatedAt.GetTime().After(end)) ||
+				(direction == "desc" && pr.CreatedAt.GetTime().Before(start))
+			if afterRange {
+				break forEachPage
+			}
+			if len(authors) > 0 && !authors[pr.GetUser().GetLogin()] {
+				continue
+			}
+			result = append(result, pr)
+		}
+		if page == resp.NextPage || page >= resp.LastPage {
+			// no more pages (GH sets 'resp.LastPage' to 0 on the last page, but I
+			// don't know if this is guaranteed. nextPage >= resp.LastPage seemed
+			// safer)
+			break
+		}
+		page = resp.NextPage
+	}
+	return result
+}
+
+// getPRs just retrieves the specific, given PRs from GitHub, instead of listing
+// all PRs and retrieving the ones that match given criteria. Called from main()
+// if '--pr-numbers' is set.
+func getPRs(client *github.Client, prs []int) []*github.PullRequest {
+	var result []*github.PullRequest
+	for _, n := range prs {
+		pr, resp, err := client.PullRequests.Get(context.Background(), "pachyderm", "pachyderm", n)
```

Comment:
It would be best to create a context in main.main with signal.NotifyContext(..., os.Interrupt) and pass it through to here.  Then Control-C runs the error handler on hanging HTTP calls, which might be helpful for debugging.